### ISSUE TITLE
Unicode characters seem to be rejected with JSON_TOKENER_STRICT

### ISSUE THREAD

**Issue by fbrosda** [json-c/json-c#875]
**Describe the bug**
When trying to parse a string, which contains unicode characters (e.g. umlauts) and setting the flag `JSON_TOKENER_STRICT`, parsing is not successful.  The error code is `json_tokener_error_parse_string`.  

**Steps To Reproduce**
A minimal example:

```c
#include <stdio.h>
#include <json.h>

int main() {
  char* test = "\"\u00e4 Ü \u20AC ß\"";
  printf("Test-String: %s\n", test);

  struct json_tokener *tok = json_tokener_new();
  json_tokener_set_flags(tok, JSON_TOKENER_STRICT);

  struct json_object *obj = json_tokener_parse_ex(tok, test, -1);

  if(tok->err != json_tokener_success)
    printf("Error-Code: %u\n", tok->err);
  
  return 0;
}
```

This prints:

```
Test-String: "ä Ü € ß"
Error-Code: 12
```

If the line with `json_tokener_set_flags` it will print no error code.

**Version and Platform**
- json-c version: json-c-0.18-1
- OS: Arch Linux

**Comment by hawicz** [comment]
This seems to work fine for me using the json-c-0.18 branch and building on both WSL and Amazon Linux.

Does Arch Linux patch json-c to cause it to behave differently?  (probably not)

Or, is your compiler doing something odd when parsing those unicode escapes?  Perhaps do a `objdump -s -j .rodata a.out` and see if you get the expected UTF-8 encoding:
```
...
 420020 22c3a420 c39c20e2 82ac20c3 9f220000  ".. .. ... .."..
...
```
i.e. 0xc3 0xa4, followed by 0xc3 0x9c, followed by 0xe2 0x82 0xac, followed by 0xc3 0x9f

It would be useful to know exactly where in the string it fails, please print `json_tokener_get_parse_end(tok)`

**Comment by fbrosda** [comment]
> Does Arch Linux patch json-c to cause it to behave differently? (probably not)

I don't think so.  Looking at the [PKGBUILD](https://gitlab.archlinux.org/archlinux/packaging/packages/json-c/-/blob/main/PKGBUILD?ref_type=heads) there seems to be nothing special.

> i.e. 0xc3 0xa4, followed by 0xc3 0x9c, followed by 0xe2 0x82 0xac, followed by 0xc3 0x9f

```
$ objdump -s -j .rodata a.out 

a.out:     file format elf64-x86-64

Contents of section .rodata:
 2000 01000200 22c3a420 c39c20e2 82ac20c3  ....".. .. ... .
 2010 9f220054 6573742d 53747269 6e673a20  .".Test-String: 
 2020 25730a00 4572726f 722d436f 64653a20  %s..Error-Code: 
 2030 25750a00 456e642d 546f6b65 6e3a2025  %u..End-Token: %
 2040 6c750a00                             lu..            

```

For me, this looks fine, too.

> It would be useful to know exactly where in the string it fails, please print json_tokener_get_parse_end(tok)

Adding `printf("End-Token: %lu\n", json_tokener_get_parse_end(tok));` to the code above, the printed value is 1. 

After this I build the library from source to test this.  `make test` succeeded so I believe the lib was build correctly.  For me there is still an error when using the 0.18 tag, but it works with the 0.17 tag.

After this I did a quick `git bisect` to find the first commit, where the error appears and I end up at 

```
6bfab90c87c27e79eae28b775938756d2fdaf6c9 is the first bad commit
6bfab90 (origin/disallow-control-chars) Issue #867: disallow control characters in strict mode., Mon Sep 2 09:43 <Eric Hawicz>
 json_tokener.c     |  6 ++++++
 tests/test_parse.c | 40 +++++++++++++++++++++++++++++++++++++++-
 2 files changed, 45 insertions(+), 1 deletion(-)
```

Does this help anything?

**Comment by hawicz** [comment]
That sort of makes sense, that commit added this check:

```
... if ((tok->flags & JSON_TOKENER_STRICT) && c <= 0x1f) ...
```

but I don't see how any character in the input would match.

**Comment by fbrosda** [comment]
Some more testing and some googling later, it looks like `char` can be signed or unsigned and it is up to the implementation what to use.

Here a small test:

```c
#include <stdio.h>

int main() {
  const char* c  = "ä";
  printf("Char-Test: %d\n", *c);
  
  return 0;
}

```

For gcc you can force signed or unsigned chars via `-fsigned-char` and `-funsigned-char` and you will get different outputs:

```
$ gcc -fsigned-char char_test.c && ./a.out
Char-Test: -61
$ gcc -funsigned-char char_test.c && ./a.out
Char-Test: 195
```

Based on what is written in https://stackoverflow.com/a/2054959 the default for gcc seems to be signed chars.  Maybe it is enough to check if `c >= 0` in the above test for control characters?

**Comment by hawicz** [comment]
Good find!  I'll do a bit more poking around to make sure I understand why it's worked everywhere I've tried it, then it can probably be fixed with a cast to get the desired behavior.  I wonder what else is broken...

**Comment by hawicz** [comment]
Oops, my previous tests were not failing before because I made a mistake while linking and ended up using an earlier version of libjson-c.so.  🤦
Once I fixed that, I am able to reproduce the problem with both your sample program, and with the json_parse too, e.g.l:
```
python3 -c 'print("\"\u00e4\"")' | ./build/apps/json_parse -s /dev/stdin
```
Casting to unsigned char fixes the check.  I'll make that change, and add some tests.
